File: leetcode_learning/code_cpp/34.cpp
class Solution { 
public:
    int binarySearchLeft(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;
        int ans = 0;
        while (left <= right)
        {
            int mid = (left + right) / 2;
            if (nums[mid] >= target)
            {
                right = mid - 1;
                ans = mid;
            }
            else
                left = mid + 1;
        }
        return ans;
    }

    int binarySearchRight(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;
        int ans = nums.size();
        while (left <= right)
        {
            int mid = (left + right) / 2;
            if (nums[mid] > target)
            {
                right = mid - 1;
                ans = mid;
            }
            else
                left = mid + 1;
        }
        return ans;
    }

    vector<int> searchRange(vector<int>& nums, int target)
    {
        int leftIdx = binarySearchLeft(nums, target);
        //cout<<leftIdx<<endl;
        int rightIdx = binarySearchRight(nums, target)-1;
        //cout<<rightIdx<<endl;
        if (leftIdx <= rightIdx && rightIdx < nums.size() && nums[leftIdx] == target && nums[rightIdx] == target)
            return vector<int>{leftIdx, rightIdx};
        return vector<int>{-1, -1};
    }
};

File: leetcode_learning/code_py/34.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        first = -1
        last = -1
        left = 0
        right = len(nums) - 1

        # 寻找第一个
        while(left<=right):
            mid = (left + right) // 2
            if nums[mid]==target:
                first = mid
                right = mid - 1 # 继续向前找
            elif nums[mid]>target:
                right = mid - 1
            else:
                left = mid + 1
        
        left = 0
        right = len(nums) - 1
        # 寻找最后一个
        while(left<=right):
            mid = (left + right) // 2
            if nums[mid]==target:
                last = mid
                left = mid + 1 # 继续向后找
            elif nums[mid]>target:
                right = mid - 1
            else:
                left = mid + 1
        
        return [first, last]
